### PR TITLE
EPMRPP-37506 || Integrations refactoring dao

### PR DESCRIPTION
**Pull request by Yumfriez**

### CODE REVIEW

**Comment by pbortnik** [comment]
Also, as we have stared writing unit test - all new code should be covered.

File: src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java [hunk 1]
```diff
@@ -45,9 +48,29 @@
 	 */
 	List<Integration> findAllByProjectId(Long projectId);
 
+	List<Integration> findAllByProjectIdAndType(Long projectId, IntegrationType integrationType);
+
+	void deleteAllByTypeIntegrationGroup(IntegrationGroupEnum integrationGroup);
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java]: There is an IntegrationType variable with name - "type" in the Integration class. IntegrationGroup variable is in the IntegrationType class. So to use JPA "delete" method i have to follow the rules of the JPA method name strategy

File: src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java [hunk 2]
```diff
@@ -45,9 +48,29 @@
...
 	@Query(value = "SELECT i.id, i.enabled, i.project_id, i.creation_date, i.params, i.type, 0 as clazz_ FROM integration i"
 			+ " WHERE (params->'params'->>'url' = :url AND params->'params'->>'project' = :btsProject"
 			+ " AND i.project_id = :projectId) LIMIT 1", nativeQuery = true)
-	Optional<Integration> findByUrlAndBtsProjectAndProjectId(@Param("url") String url, @Param("btsProject") String btsProject,
+	Optional<Integration> selectProjectBtsIntegrationByUrlAndLinkedProject(@Param("url") String url, @Param("btsProject") String btsProject,
 			@Param("projectId") Long projectId);
+
+	@Query(value = "SELECT i.id, i.enabled, i.project_id, i.creation_date, i.params, i.type FROM integration i "
+			+ " WHERE params->'params'->>'url' = :url AND i.params->'params'->>'project' = :projectName AND i.project_id IS NULL", nativeQuery = true)
+	Optional<Integration> selectGlobalBtsIntegrationByUrlAndLinkedProject(@Param("url") String url,
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java]: Let it be 'findGlobalBtsByUrlAnd**Linked**Project', cause we search for BTS by its BTS-project, not by ReportPortal Project.

File: src/main/java/com/epam/ta/reportportal/dao/IntegrationRepositoryCustom.java [hunk 3]
```diff
@@ -35,4 +36,8 @@
 
 	Optional<ActiveDirectoryConfig> findActiveDirectory(boolean enabled);
 
+	Optional<Integration> getGlobalIntegrationById(Long integrationId);
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/IntegrationRepositoryCustom.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/IntegrationRepositoryCustom.java [hunk 4]
```diff
@@ -35,4 +36,8 @@
...
+
+	List<Integration> getAllMissedGlobalIntegrations(List<Long> integrationTypeIds);
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/IntegrationRepositoryCustom.java]: done

File: src/main/java/com/epam/ta/reportportal/entity/ServerSettingsEnum.java [hunk 5]
```diff
@@ -17,26 +17,25 @@
...
-	HOST(EMAIL_CONFIG_PREFIX + "host"),
-	PORT(EMAIL_CONFIG_PREFIX + "port"),
-	PROTOCOL(EMAIL_CONFIG_PREFIX + "protocol"),
-	AUTH_ENABLED(EMAIL_CONFIG_PREFIX + "auth_enabled"),
-	STAR_TLS_ENABLED(EMAIL_CONFIG_PREFIX + "star_tls_enabled"),
-	SSL_ENABLED(EMAIL_CONFIG_PREFIX + "ssl_enabled"),
-	USERNAME(EMAIL_CONFIG_PREFIX + "username"),
-	PASSWORD(EMAIL_CONFIG_PREFIX + "password"),
-	FROM(EMAIL_CONFIG_PREFIX + "from"),
+	HOST("host"),
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/entity/ServerSettingsEnum.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java [hunk 6]
```diff
@@ -45,9 +48,29 @@
...
+	void deleteAllByTypeIntegrationGroup(IntegrationGroupEnum integrationGroup);
+
+	@Query(value = "SELECT i FROM Integration i WHERE i.project IS NULL AND i.type = :integrationType")
+	List<Integration> getAllGlobalIntegrationsByType(@Param("integrationType") IntegrationType integrationType);
+
 	@Query(value = "SELECT i.id, i.enabled, i.project_id, i.creation_date, i.params, i.type, 0 as clazz_ FROM integration i"
 			+ " WHERE (params->'params'->>'url' = :url AND params->'params'->>'project' = :btsProject"
 			+ " AND i.project_id = :projectId) LIMIT 1", nativeQuery = true)
-	Optional<Integration> findByUrlAndBtsProjectAndProjectId(@Param("url") String url, @Param("btsProject") String btsProject,
+	Optional<Integration> selectProjectBtsIntegrationByUrlAndLinkedProject(@Param("url") String url, @Param("btsProject") String btsProject,
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java [hunk 7]
```diff
@@ -45,9 +48,29 @@
...
 			@Param("projectId") Long projectId);
+
+	@Query(value = "SELECT i.id, i.enabled, i.project_id, i.creation_date, i.params, i.type FROM integration i "
+			+ " WHERE params->'params'->>'url' = :url AND i.params->'params'->>'project' = :projectName AND i.project_id IS NULL", nativeQuery = true)
+	Optional<Integration> selectGlobalBtsIntegrationByUrlAndLinkedProject(@Param("url") String url,
+			@Param("projectName") String projectName);
+
+	@Modifying
+	@Query(value = "UPDATE integration SET enabled = :enabled WHERE id = :integrationId", nativeQuery = true)
+	void updateIntegrationEnabledState(@Param("enabled") boolean enabled, @Param("integrationId") Long integrationId);
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java [hunk 8]
```diff
@@ -45,9 +48,29 @@
...
+
+	@Modifying
+	@Query(value = "UPDATE integration SET enabled = :enabled WHERE type = :integrationTypeId", nativeQuery = true)
+	void updateIntegrationGroupEnabledState(@Param("enabled") boolean enabled, @Param("integrationTypeId") Long integrationTypeId);
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java [hunk 9]
```diff
@@ -45,9 +48,29 @@
 	 */
 	List<Integration> findAllByProjectId(Long projectId);
 
+	List<Integration> findAllByProjectIdAndType(Long projectId, IntegrationType integrationType);
+
+	void deleteAllByTypeIntegrationGroup(IntegrationGroupEnum integrationGroup);
+
+	@Query(value = "SELECT i FROM Integration i WHERE i.project IS NULL AND i.type = :integrationType")
+	List<Integration> getAllGlobalIntegrationsByType(@Param("integrationType") IntegrationType integrationType);
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java]: done

File: src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java [hunk 10]
```diff
@@ -38,16 +40,80 @@
...
+	 * @param url        Bug Tracking System url
+	 * @param btsProject Bug Tracking System project name
+	 * @param projectId  {@link com.epam.ta.reportportal.entity.project.Project#id}
+	 * @return The {@link Integration} wrapped in the {@link Optional}
+	 */
 	@Query(value = "SELECT i.id, i.enabled, i.project_id, i.creation_date, i.params, i.type, 0 as clazz_ FROM integration i"
 			+ " WHERE (params->'params'->>'url' = :url AND params->'params'->>'project' = :btsProject"
 			+ " AND i.project_id = :projectId) LIMIT 1", nativeQuery = true)
-	Optional<Integration> findByUrlAndBtsProjectAndProjectId(@Param("url") String url, @Param("btsProject") String btsProject,
+	Optional<Integration> findProjectBtsByUrlAndLinkedProject(@Param("url") String url, @Param("btsProject") String btsProject,
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/dao/IntegrationRepository.java]: If an integration is a abstraction for all supported integrations, why do we need an explicit method for Bts?
Maybe it is better make bts type as a parameter and add it to the sql query ?

**Comment by pbortnik** [comment]
By the way, coverage is still decreased

File: src/test/java/com/epam/ta/reportportal/dao/suite/IntegrationRepositoryTest.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+		IntegrationType integrationType = integrationTypeRepository.findById(JIRA_INTEGRATION_TYPE_ID).get();
+
+		List<Integration> integrations = integrationRepository.findAllByProjectIdAndType(DEFAULT_PERSONAL_PROJECT_ID, integrationType);
+
+		Assert.assertNotNull(integrations);
+		Assert.assertEquals(2, integrations.size());
+	}
+
+	public void test() {
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/test/java/com/epam/ta/reportportal/dao/suite/IntegrationRepositoryTest.java]: a missprint

File: src/test/java/com/epam/ta/reportportal/dao/suite/IntegrationRepositoryTest.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+		IntegrationType integrationType = integrationTypeRepository.findById(JIRA_INTEGRATION_TYPE_ID).get();
+
+		integrationRepository.deleteAllByTypeId(integrationType.getId());
+
+		Assert.assertTrue(CollectionUtils.isEmpty(integrationRepository.findAllByProjectIdAndType(DEFAULT_PERSONAL_PROJECT_ID,
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/test/java/com/epam/ta/reportportal/dao/suite/IntegrationRepositoryTest.java]: done

File: src/test/java/com/epam/ta/reportportal/dao/suite/IntegrationTypeRepositoryTest.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+	private final static String JIRA_INTEGRATION_TYPE_NAME = "JIRA";
+	private final static String WRONG_INTEGRATION_TYPE_NAME = "WRONG";
+
+	@Autowired
+	private IntegrationTypeRepository integrationTypeRepository;
+
+	@Test
+	public void shouldFindWhenNameExists() {
+
+		Assert.assertTrue(integrationTypeRepository.findByName(JIRA_INTEGRATION_TYPE_NAME).isPresent());
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/test/java/com/epam/ta/reportportal/dao/suite/IntegrationTypeRepositoryTest.java]: done

**Pull request merged**